runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -212,12 +212,15 @@ local _bufs_waiting_to_update = setmetatable({}, bufnr_and_client_cacher_mt)
 ---
 ---@param diagnostics Diagnostic[]
 ---@return table<number, Diagnostic[]>
-local _diagnostic_lines = function(diagnostics)
+local _diagnostic_lines = function(diagnostics, bufnr)
   if not diagnostics then return end
 
+  local buf_line_count = vim.api.nvim_buf_line_count(bufnr)
   local diagnostics_by_line = {}
   for _, diagnostic in ipairs(diagnostics) do
     local start = diagnostic.range.start
+    -- Account for servers that place diagnostics on terminal newline
+    start.line = math.min(start.line, buf_line_count - 1)
```

Comment:
What do you think of moving this into the loop here https://github.com/neovim/neovim/blob/c1fbc2ddf15b2f44b615f90b2511349ab974cb83/runtime/lua/vim/lsp/diagnostic.lua#L267-L271 so that there's only a single place where we mutate the diagnostics?

Reply:
Yeah, I initially did this to save the cost of getting the line count of the buffer, but I realize now it's a super bad idea for the line count in the line cache and the per buffer cache to not match